resolver/custom_dns_resolver.go

```diff
@@ -120,26 +144,113 @@ func (r *CustomDNSResolver) processRequest(request *model.Request) *model.Respon
+	return response, err
+}
+
+func (r *CustomDNSResolver) processDNSEntry(
+	ctx context.Context,
+	request *model.Request,
+	resolvedCnames []string,
+	question dns.Question,
+	entry dns.RR,
+) ([]dns.RR, error) {
+	switch v := entry.(type) {
+	case *dns.A:
+		return r.processIP(v.A, question)
+	case *dns.AAAA:
+		return r.processIP(v.AAAA, question)
+	case *dns.CNAME:
+		return r.processCNAME(ctx, request, *v, resolvedCnames, question)
+	}
+
+	return nil, fmt.Errorf("unsupported type %T", entry)
 }
 
 // Resolve uses internal mapping to resolve the query
 func (r *CustomDNSResolver) Resolve(ctx context.Context, request *model.Request) (*model.Response, error) {
-	logger := log.WithPrefix(request.Log, "custom_dns_resolver")
-
 	reverseResp := r.handleReverseDNS(request)
 	if reverseResp != nil {
 		return reverseResp, nil
 	}
 
-	if len(r.mapping) > 0 {
-		resp := r.processRequest(request)
-		if resp != nil {
-			return resp, nil
+	resp, err := r.processRequest(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+func (r *CustomDNSResolver) processIP(ip net.IP, question dns.Question) (result []dns.RR, err error) {
+	result = make([]dns.RR, 0)
+
+	if isSupportedType(ip, question) {
+		rr, err := r.createAnswerFromQuestion(question, ip, r.cfg.CustomTTL.SecondsU32())
+		if err != nil {
+			return nil, err
 		}
+
+		result = append(result, rr)
 	}
 
-	logger.WithField("next_resolver", Name(r.next)).Trace("go to next resolver")
+	return result, nil
+}
+
+func (r *CustomDNSResolver) processCNAME(
+	ctx context.Context,
+	request *model.Request,
+	targetCname dns.CNAME,
+	resolvedCnames []string,
+	question dns.Question,
+) (result []dns.RR, err error) {
+	cname := new(dns.CNAME)
+	ttl := r.cfg.CustomTTL.SecondsU32()
+	cname.Hdr = dns.RR_Header{Class: dns.ClassINET, Ttl: ttl, Rrtype: dns.TypeCNAME, Name: question.Name}
+	cname.Target = dns.Fqdn(targetCname.Target)
+	result = append(result, cname)
+
+	targetWithoutDot := strings.TrimSuffix(targetCname.Target, ".")
+
+	if slices.Contains(resolvedCnames, targetWithoutDot) {
+		return nil, fmt.Errorf("circular reference detected for domain %s", targetWithoutDot)
+	}
+
+	clientIP := request.ClientIP.String()
+	clientID := request.RequestClientID
+	aRequest := newRequestWithClientID(targetWithoutDot, dns.Type(dns.TypeA), clientIP, clientID)
+	aaaaRequest := newRequestWithClientID(targetWithoutDot, dns.Type(dns.TypeAAAA), clientIP, clientID)
+
+	// Resolve target recursively
+	targetResp, resolvedCnames, err := r.processRequestTrackingCnames(ctx, aRequest, append(resolvedCnames, targetWithoutDot))
+	if err != nil {
+		return nil, err
+	}
+
+	result = append(result, targetResp.Res.Answer...)
+
+	// Resolve ipv6 target recursively
+	targetResp, resolvedCnames, err = r.processRequestTrackingCnames(ctx, aaaaRequest, append(resolvedCnames, targetWithoutDot))
+	if err != nil {
+		return nil, err
+	}
+
+	result = append(result, targetResp.Res.Answer...)
```

Comment:
Sorry to tack this on last minute, but the logic needs to be adjusted. My bad for not realizing earlier and suggesting the AAAA thing.

Based on [RFC-1034 Section 3.6.2](https://www.rfc-editor.org/rfc/rfc1034#section-3.6.2):
- we don't need to do both A and AAAA, but should use what the question asks for. Looking at what `RewriterResolver.rewriteRequest` does should be a good starting point
- if the question is for a CNAME, then we must not recurse but send the CNAME itself
- a CNAME RR must be the only one for a domain

For the last point I'm not sure what's the best thing to do if the user's config doesn't respect that. Here's a couple options:
- error in `UnmarshalYAML`, the user will notice immediately since they're presumably editing and restarting blocky
- log an error and return a ServFail response. EDE, if enabled, could point the user in the right direction.
- discard the extra RRs and just return the result from the CNAME query

Reply:
The other changes requested have been pushed. I'll start taking a look at this one now, as it seems like a bigger change than the others

Reply:
Would it be correct to say that a domain could have multiple CNAME RRs associated but that all RRs should be CNAMEs if any one is a CNAME? 

I believe that it is valid to have 2 CNAMEs for a single domain, pointing to different targets.

Reply:
Looks like my understanding is wrong. Only a single CNAME value is valid for a given domain. I'll add a check to the UnmarshalYAML function so that users are immediately forced to correct the error rather than blindly ignoring configuration.

Reply:
> Only a single CNAME value is valid for a given domain

Yeah that's my understanding too!